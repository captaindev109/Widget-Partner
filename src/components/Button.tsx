import styled, { DefaultTheme } from 'styled-components';
import { FC } from 'react';
import type { BasicBool, ButtonProps, ButtonVariant } from '@/types/types';
import Card from '@/generics/Card';
import { CreateSVGComponent } from './SVGIcons/Index';

interface VariantSpecificStyling {
  variant?: 'primary' | 'outline' | 'transparent' | 'icon';
  fluid?: boolean;
  theme: DefaultTheme;
}

const getVariantSpecificStyling = ({ theme, variant }: VariantSpecificStyling) => {
  switch (variant) {
    case 'primary':
      return `
        cursor: pointer;
        background-color: ${theme.color.gray900};
        color: ${theme.color.gray100};
        border:none
        cursor: pointer;
        &:hover {
          cursor: pointer;
          background-color: ${theme.color.gray800};
        }
        &:disabled {
          cursor: default;
          background-color: ${theme.color.gray300};
          color: ${theme.color.gray500};
        }
      `;
    case 'outline':
      return `
        background-color: ${theme.palette.common.transparent};
        color: ${theme.color.gray900};
        border: 1px solid ${theme.color.gray900};
        &:hover {
          cursor: pointer;
          border: 1px solid ${theme.color.gray700};
        }
        &:disabled {
          cursor: default;
          background-color: ${theme.palette.common.transparent};
          border: 1px solid ${theme.color.gray500};
          color: ${theme.color.gray500};
        }
      `;
    case 'transparent':
      return `
        background-color: ${theme.palette.common.transparent};
        color: ${theme.color.gray900};
        &:hover {
          cursor: pointer;
          background-color: ${theme.color.gray300};
        }
        &:disabled {
          cursor: default;
          background-color: ${theme.palette.common.transparent};
          color: ${theme.color.gray500};
          & > div > div > svg {
            color: ${theme.color.gray500};
            fill: ${theme.color.gray500};
          }
        }
      `;
    case 'icon':
      return `
        background: none;
        cursor: pointer;
        padding: 0;
        height: 32px;
        width: 32px;
        z-index: 1;
      `;
    default:
      return ``;
  }
};

const getIconColor = ({
  variant,
  disabled,
  theme,
}: {
  variant?: ButtonVariant;
  disabled?: BasicBool;
  theme: DefaultTheme;
}) => {
  if (disabled) return theme.color.gray500;
  if (variant === 'primary') return theme.color.gray100;
  return theme.color.gray900;
};

const StyledButton = styled.button<VariantSpecificStyling>`
  display: flex;
  flex: 1;
  width: ${(props) => (props.fluid ? 'auto' : '100%')};
  align-items: center;
  justify-content: center;
  padding: 0px 32px;
  margin: 0;
  height: 48px;
  border-radius: 8px;
  border: none;
  pointer: cursor;
  font-family: ${(props) => props.theme.fonts.FAVORIT};
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  ${getVariantSpecificStyling}
`;

const ContentWrapper = styled(Card)`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const IconWrapper = styled(Card)<{ disabled?: BasicBool; variant?: ButtonVariant }>`
  ${(props) => props.variant !== 'icon' && `padding-left: 16px;`}
  display: flex;
  align-items: center;
  justify-content: center;
  & > svg > * {
    fill: ${getIconColor};
  }
`;

const Button: FC<ButtonProps> = ({ children, onClick, icon, ...rest }) => {
  return (
    <StyledButton onClick={onClick} {...rest}>
      <ContentWrapper>
        {children}
        {icon && (
          <IconWrapper variant={rest.variant} disabled={rest.disabled}>
            {CreateSVGComponent(icon.iconName, icon.config)}
          </IconWrapper>
        )}
      </ContentWrapper>
    </StyledButton>
  );
};

Button.defaultProps = {
  type: 'button',
  disabled: false,
  variant: 'primary',
};

export default Button;
