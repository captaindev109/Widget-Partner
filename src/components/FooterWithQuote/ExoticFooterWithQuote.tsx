import { FC } from 'react';
import useStore from '@/hooks/useStore';
import FooterWithQuote from '.';

// A basic Interface or type is required for every each component
export interface ExoticFooterWithQuoteProps {
  children: React.ReactNode;
}

//The main component with dynamic rendering and state
const ExoticFooterWithQuote: FC<ExoticFooterWithQuoteProps> = ({ children }) => {
  const { cartCostSelector } = useStore();
  return <FooterWithQuote price={cartCostSelector}>{children}</FooterWithQuote>;
};

export default ExoticFooterWithQuote;
