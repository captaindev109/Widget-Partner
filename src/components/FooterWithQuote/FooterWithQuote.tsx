import { FC } from 'react';
import Text from '@/generics/Text';
import Card from '@/generics/Card';
import { MonetaryAmountV2 } from '@/types/schemaTypes';

// A basic Interface or type is required for every each component
export interface FooterWithQuoteProps {
  price: MonetaryAmountV2;
  children: React.ReactNode;
}

//The main component with dynamic rendering and state
const FooterWithQuote: FC<FooterWithQuoteProps> = ({ price, children }) => {
  return (
    <Card display="flex" alignItems="center" data-testid="FooterWithQuote">
      <Card flex="auto">
        <Text variant="Body2">
          {price.amount} {price.currency}/mnd
        </Text>
        <Text variant="Caption" color="gray700">
          Avsluta n√§r du vill
        </Text>
      </Card>
      <Card>{children}</Card>
    </Card>
  );
};

export default FooterWithQuote;
