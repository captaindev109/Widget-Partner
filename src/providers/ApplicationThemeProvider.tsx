import { ThemeProvider, DefaultTheme } from 'styled-components';
import GlobalStyles from '@/styles/Global';
import React from 'react';
import { ChildrenReturnType } from '@/types/types';

interface ApplicationThemeProviderProps {
  children?: ChildrenReturnType;
  theme: DefaultTheme;
}

const ApplicationThemeProvider = ({ children, theme }: ApplicationThemeProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};

export default ApplicationThemeProvider;
