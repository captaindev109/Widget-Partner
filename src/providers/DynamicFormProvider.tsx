import React from 'react';
import { ChildrenReturnType } from '@/types/types';
import { FormProvider, useForm } from 'react-hook-form';

interface DynamicFormProviderProps {
  children?: ChildrenReturnType;
}

const useDynamicRootForm = () => {
  const methods = useForm({ mode: 'onChange' });

  return { methods };
};

export const DynamicFormProvider = ({ children }: DynamicFormProviderProps) => {
  const { methods } = useDynamicRootForm();

  return <FormProvider {...methods}>{children}</FormProvider>;
};
