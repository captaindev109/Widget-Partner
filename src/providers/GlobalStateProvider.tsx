import { mainMachine } from '@/machines/courseMachine';
import headerMachine from '@/machines/headerMachine';
import footerMachine from '@/machines/footerMachine';
import React, { createContext } from 'react';
import { useInterpret } from '@xstate/react';
import { ActorRefFrom } from 'xstate';
import { storeMachine } from '@/machines/storeMachine';

interface GlobalStateContextProps {
  mainService: ActorRefFrom<typeof mainMachine>;
  headerService: ActorRefFrom<typeof headerMachine>;
  footerService: ActorRefFrom<typeof footerMachine>;
  storeService: ActorRefFrom<typeof storeMachine>;
}

export const GlobalStateContext = createContext({} as GlobalStateContextProps);

interface GlobalStateProviderProps {
  children?: React.ReactNode;
}

const useDevTool = process.env.NODE_ENV === 'development';

export const GlobalStateProvider = ({ children }: GlobalStateProviderProps) => {
  const services = {
    mainService: useInterpret(mainMachine, {
      devTools: useDevTool,
    }),
    headerService: useInterpret(headerMachine, {
      devTools: useDevTool,
    }),
    footerService: useInterpret(footerMachine, {
      devTools: useDevTool,
    }),
    storeService: useInterpret(storeMachine, {
      devTools: useDevTool,
    }),
  };

  return <GlobalStateContext.Provider value={services}>{children}</GlobalStateContext.Provider>;
};
