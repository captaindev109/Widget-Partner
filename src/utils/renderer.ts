import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
