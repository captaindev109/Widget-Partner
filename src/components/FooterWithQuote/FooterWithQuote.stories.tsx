import React from 'react';
import { Meta, Story } from '@storybook/react';
import styled from 'styled-components';
import { decorators } from '@/../.storybook/preview';
import Wrapper from '@/generics/Wrapper';
import FooterWithQuote from '.';
import { FooterWithQuoteProps } from './FooterWithQuote';
import Button from '../Button';

// A wrapper to position the component
const StoryWrapper = styled(Wrapper)``;

// Name the folder to either generics or components
const meta: Meta = {
  title: 'Components/FooterWithQuote',
  component: FooterWithQuote,
  decorators,
};

export default meta;

// Showcasing the component
const Template: Story<FooterWithQuoteProps> = (args) => (
  <StoryWrapper>
    <FooterWithQuote {...args}>
      <Button icon={{ iconName: 'BankIdIcon' }}>Button</Button>
    </FooterWithQuote>
    <FooterWithQuote {...args}>
      <Button icon={{ iconName: 'BankIdIcon' }}>Button with icon</Button>
    </FooterWithQuote>
  </StoryWrapper>
);

export const DefaultFooterWithQuote = Template.bind({});

DefaultFooterWithQuote.args = {
  price: { amount: '189', currency: 'kr' },
};
