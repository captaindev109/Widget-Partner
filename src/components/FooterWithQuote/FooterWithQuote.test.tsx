/**
 *
 *  ********** Unit & Snapshot Testing *********
 *
 * Unit test is specifiic for each functionliaty of the component
 * Snapshot test is to track Over All changes in the code since the first snap
 **/

import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import FooterWithQuote from './FooterWithQuote';
import defaultTheme from '@/theme/theme';
import ApplicationThemeProvider from '@/providers/ApplicationThemeProvider';
import { GlobalStateProvider } from '@/providers/GlobalStateProvider';
import { DynamicFormProvider } from '@/providers/DynamicFormProvider';
import Button from '../Button';

const MockComponet: React.FC = () => {
  return (
    <GlobalStateProvider>
      <ApplicationThemeProvider theme={defaultTheme}>
        <DynamicFormProvider>
          <FooterWithQuote price={{ amount: '100', currency: 'kr' }}>
            <Button>Slutfor kop</Button>
          </FooterWithQuote>
        </DynamicFormProvider>
      </ApplicationThemeProvider>
    </GlobalStateProvider>
  );
};

// describe function is a better way to organize our test cases and describe which component we are testing
describe('<FooterWithQuote /> Test Cases', () => {
  // this test is a must for each component, any rendering issues with the component can be found
  test('it renders without crashing, all is well', () => {
    // render basically prints and exposes the component in nodejs environemt, it returns an object...
    render(<MockComponet />);
  });

  it('Snapshot test should pass', () => {
    const snap = renderer.create(<MockComponet />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
