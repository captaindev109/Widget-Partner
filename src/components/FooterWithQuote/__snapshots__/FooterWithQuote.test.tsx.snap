// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FooterWithQuote /> Test Cases Snapshot test should pass 1`] = `
<div
  className="Card-sc-fb64gk-0 hskMYt"
  data-testid="FooterWithQuote"
  display="flex"
>
  <div
    className="Card-sc-fb64gk-0 fjEoSy"
  >
    <p
      className="Text__StyledText-sc-1jfsuh1-0 iWpXoI"
      color="gray900"
    >
      100
       
      kr
      /mnd
    </p>
    <p
      className="Text__StyledText-sc-1jfsuh1-0 eTNRxW"
      color="gray700"
    >
      Avsluta n√§r du vill
    </p>
  </div>
  <div
    className="Card-sc-fb64gk-0 leCJCh"
  >
    <button
      className="Button__StyledButton-sc-1drbzk-0 ihwfYJ"
      disabled={false}
      type="button"
    >
      <div
        className="Card-sc-fb64gk-0 Button__ContentWrapper-sc-1drbzk-1 leCJCh jowrmu"
      >
        Slutfor kop
      </div>
    </button>
  </div>
</div>
`;
